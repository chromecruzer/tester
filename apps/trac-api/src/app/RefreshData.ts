import {
  ImportIntoConsignments,
  ImportIntoJumpTable,
  NotificationsServer,
  PostgresClient,
  SearchTable,
  tryCatch
} from "@trac/postgresql";
import {uploadHeaders} from "@trac/datatypes";
import {StatusCodes} from "http-status-codes";
import UploadManager from "./datastore/UploadManager";
import ScheduleBatches from "./emails/ScheduleBatches";

export default class RefreshData {
  private search: SearchTable;
  private consignments: ImportIntoConsignments;
  private salesJump: ImportIntoJumpTable;

  constructor(private schemaName, public uploadManager: UploadManager, private scheduler: ScheduleBatches,
              private ns = null as (NotificationsServer | null)) {
    this.search = new SearchTable(schemaName);
    this.consignments = new ImportIntoConsignments(schemaName, 'consignments', 'iol',
      'products');
    this.salesJump = new ImportIntoJumpTable(schemaName, 'salesjumptable', 'salesreps',
      'salesmappings');
  }

  private async clear(client) {
    if (this.ns) {
      this.ns.notify('INFO', `clear old data`,
        'removing search table', ['iolUpload'])
    }
    console.log('removing search table');
    await this.search.removeTable(client);
    if (this.ns) {
      this.ns.notify('INFO', `clear old data`,
        'removing consignments table', ['iolUpload'])
    }
    console.log('removing consignments table');
    await this.consignments.removeTable(client);
    if (this.ns) {
      this.ns.notify('INFO', `clear old data`,
        'removing sales jumps table', ['iolUpload'])
    }
    console.log('removing sales jumps table');
    await this.salesJump.removeTable(client);
  }

  private async refill(client) {
    if (this.ns) {
      this.ns.notify('INFO', `re-creating old data`,
        're-filling consignments table', ['iolUpload'])
    }
    console.log('re-filling consignments table');
    await this.consignments.into(client);
    await this.consignments.removeExcluded( client);
    if (this.ns) {
      this.ns.notify('INFO', `re-creating  old data`,
        're-filling sales jumps table', ['iolUpload'])
    }
    console.log('re-filling sales jumps table');
    await this.salesJump.fill(client);
    if (this.ns) {
      this.ns.notify('INFO', `re-creating old data`,
        're-filling search table', ['iolUpload'])
    }
    console.log('re-filling search table');
    await this.search.createTable(client);
    const tables = ['consignments', 'customers', 'products', 'salesreps'];
    for (const t of tables) {
      await this.search.fillTable(this.schemaName, t, client)
    }
  }

  public async refresh(client, user) {
    await this.clear(client);
    await this.refill(client);
    this.scheduler.clearMissingEmailAddresses();
    return this.uploadManager.markRefresh(user);
  }

  public refreshHttp(postgresClient: PostgresClient) {
    return tryCatch(async (req, res, next) => {
      const user = req.body[uploadHeaders.userField];
      if (this.ns) {
        this.ns.started(['iolUpload']);
      }
      const client = await postgresClient.getClient()
      await this.refresh(client, user)
        .finally(async () => {
          return postgresClient.release(client);
        });
      this.uploadManager.markRefresh(user);
      res.status(StatusCodes.OK);
      if (this.ns) {
        this.ns.completed(['iolUpload']);
      }
      res.send('OK')
    })
  }
}
