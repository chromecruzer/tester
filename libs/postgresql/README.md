# @trac-postgresql

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test @trac-postgresql` to execute the unit tests via [Jest](https://jestjs.io).
