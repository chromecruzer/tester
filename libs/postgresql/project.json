{
  "name": "postgresql",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/postgresql/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/postgresql/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/libs/postgresql"],
      "options": {
        "jestConfig": "libs/postgresql/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
