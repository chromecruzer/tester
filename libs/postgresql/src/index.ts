export * from './postgresql';
import PostgresConfigFactory, {PostgresConfig} from "./PostgresConfigFactory";
import BulkSqlWithCreate from "./BulkSqlWithCreate";
export *  from './ErrorHandler';
import ExtractIolData from "./ExtractIolData";
import PostgresClient from "./PostgresClient";
import LoadXlsxData from './LoadXlsxData';
import NotificationsServer from "./NotificationsServer";
import SearchTable from "./SearchTable";
import ImportIntoConsignments from "./ImportIntoConsignments";
import ImportIntoJumpTable from "./ImportIntoJumpTable";
import AuditStore from "./AuditStore";
import UserStore from './UserStore';

export {
  AuditStore, PostgresConfig, PostgresConfigFactory, BulkSqlWithCreate, ExtractIolData,
  PostgresClient, LoadXlsxData, NotificationsServer, SearchTable, ImportIntoConsignments, ImportIntoJumpTable,
  UserStore
};
