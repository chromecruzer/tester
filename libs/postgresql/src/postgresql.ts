import * as _ from "lodash";
import {DateTime} from "luxon";
import {PoolClient} from "pg";
import util from "util";
// eslint-disable-next-line @nrwl/nx/enforce-module-boundaries
import {
  customerContactDataMapping,
  customerDataMapping,
  dataTableNames, InputDataType,
  InputDataTypeLabel,
  InputToSqlMapping,
  iolDataMapping,
  matchCustomerContactFields,
  matchCustomerFields,
  matchIolFields,
  matchProductFields,
  matchSalesMappingFields,
  matchSalesRepFields,
  NullableString,
  productDataMapping,
  employeeRecordDm,
  salesMappingDataMapping,
  salesRepDataMapping,
  tracDateFormat, 
  UploadRecord,
  matchEmployeeFields,
  userRecordDm,
  matchUser1Fields
} from "../../datatypes/src";
import path from "path";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

export const auditTableNames = {
  notes: 'auditnotes',
  locations: 'auditlocation',
  items: 'audititem',
}



export const searchTableNamesFn = () => [
  dataTableNames.consignments,
  dataTableNames.customers,
  dataTableNames.customerContacts,
  dataTableNames.products,
  dataTableNames.salesReps];

export const sqlDate = (datum: (Date | string)) => {
  // console.log(`datum ${dump(datum)}
  // from JSDate ${dump(DateTime.fromJSDate(datum))}
  // to format ${dump(DateTime.fromJSDate(datum).toFormat(tracDateFormat))}`)
  const date = _.isDate(datum) ? datum : new Date(datum);
  return DateTime.fromJSDate(date).toFormat(tracDateFormat)
}

export function datumToSql(datum: (string | number | Date | boolean), datatype: InputDataTypeLabel) {
  if (_.isNull(datum) || _.isUndefined(datum) || datum === '') {
    return 'NULL';
  }
  let str;
  switch (datatype) {
    case 'string':
      str = String(datum);
      if (str.length === 0) {
        return null;
      }
      str = str.replace(/'/g, "''");
      str = str.replace(/\n/g, "\\n");
      return `'${str}'`;
    case 'date':
      // console.log(`date ${datum} converted ${sqlDate(datum as unknown as Date)}`)
      return `'${sqlDate(datum as Date)}'`
    case 'boolean':
      // console.log(`boolean ${datum} as ${Boolean(datum)}`)
      return Boolean(datum);
    case 'number':
      return datum;
  }
}

export function appendUuids(sql: string, uuids: string[]) {
  let skipped1st = false;
  let result = sql;
  uuids.forEach(id => {
    result += `${skipped1st ? ',' : ''}    \n'${id}'`;
    skipped1st = true;
  });
  result += ');';
  return result;
}

export const checkModified = modified => {
  // console.log(`Checking ${dump(modified)}`);
  if (modified.originalUuid !== modified.new.uuid) {
    console.error(`UUID mismatch found ${dump(modified)}`);
  }
}

export function appendBulkData(sql, bulkData: UploadRecord[],
                               columnMapping: InputToSqlMapping[],
                               createUuid = true,
                               hasUuid = false) {
  let skipped1stRow = false;
  let result = `${sql}\n`;
  bulkData.forEach((record) => {
    // console.log(`append bulk data ${dump(record)}`)
    const dataRow = [];
    if (createUuid) {
      dataRow.push('gen_random_uuid()')
    }
    if(hasUuid) {
      dataRow.push(datumToSql(record['uuid'], 'string'));
    }
    columnMapping.forEach((colMap) => {
      let datum = datumToSql(record[colMap.sqlLabel], colMap.dataType);
      if(!_.isUndefined(colMap.default) && datum === null) {
        datum = colMap.default;
      }
      dataRow.push(datum)
    });
    result += `${skipped1stRow ? ',': ''}\n(${_.join(dataRow, ', ')})`;
    skipped1stRow = true;
  });
  return result;
}

export const queryRecords = async (client: PoolClient, recordSql: string, idField: string,
                                   {
                                     uuid = null as NullableString,
                                     id = null as NullableString,
                                     ids = [] as string[],
                                   }) => {
  let sql = recordSql;
  const check = (_.isString(uuid) ? 1 : 0) + (_.isString(id) ? 1 : 0) + (_.isEmpty(ids) ? 0 : 1);
  if (check > 1) {
    throw new Error(`cannot set multiple parameters ${uuid}, ${id}, ${ids.length}`);
  }
  if (uuid) {
    sql += ` WHERE uuid='${uuid}'`;
  }
  if (id) {
    sql += ` WHERE ${idField}='${id}'`;
  }
  if (!_.isEmpty(ids)) {
    sql += ` WHERE ${idField} IN (${_.join(_.map(ids, i => `'${i}'`), ', ')})`;
  }
  sql += ';';
  // console.log(`query records sql '${sql}'`);
  return client.query(sql)
    .then(data => data.rows)
    .catch(err => {
    throw new Error(`Mistake in '${sql}' caused ${err.message}`);
  });
}

interface TransformerKeys {
  matchFields: (string | string[]);
  matchFn: (a: any, b: any) => boolean,
  mappings: InputToSqlMapping[];
  compareFieldList?: string[],
}

export const recordFromArray = (row: InputDataType[], mapping: InputToSqlMapping[]) => {
  return _.reduce(mapping, (accum, m,i) => {
    const r = row[i] || null;
    let value = _.isFunction(m.nullify) ? m.nullify(r) : r;
    if(_.has(m, 'default')) { //need to us _.has here because the default may be a boolean false
      // console.log(`checking default for ${dump(m)} with value ${value}`)
      value = value || m.default;
    }
    accum[m.sqlLabel] = value;
    return accum;
  }, {} as UploadRecord);
}
export const transformHelpers: { [id: string]: TransformerKeys } = {
  iol: {
    matchFields: 'lot',
    matchFn: matchIolFields,
    mappings: iolDataMapping,
  },
  customers: {
    matchFields: 'customer_code',
    matchFn: matchCustomerFields,
    mappings: customerDataMapping,
  },
  customer_contacts: {
    matchFields: ['full_name', 'ship_to_id'],
    matchFn: matchCustomerContactFields,
    mappings: customerContactDataMapping,
  },
  products: {
    matchFields: 'product_id',
    matchFn: matchProductFields,
    mappings: productDataMapping,
  },
  employee: {
     matchFields: 'emp_id',
     matchFn: matchEmployeeFields,          // added by shyam for testing
     mappings: employeeRecordDm,
  },

  users: {
    matchFields: 'username',
    matchFn:  matchUser1Fields,              // added by shyam for testing
    mappings: userRecordDm
  },
  salesmappings: {
    matchFields: 'customer_ship_to',
    matchFn: matchSalesMappingFields,
    mappings: salesMappingDataMapping,
    compareFieldList: ['customer_ship_to', 'postal', 'zipcode', 'primary_territory', 'primary_name', 'cos_territory',
      'cos_name', 'kae_territory', 'kae_name'],
  },
  salesreps: {
    matchFields: 'salesrep_id',
    matchFn: matchSalesRepFields,
    mappings: salesRepDataMapping,
  },
}
export const matchFieldFn = (fields, record) => {
  return _.isArray(fields) ? _.join(_.map(fields, f => record[f]), '|') : record[fields]
}
export const deepEqualFn = (c, m, compareFieldList = null) => {
  if (_.isArray(compareFieldList)) {
    return _.reduce(compareFieldList,
      (boolAccum, cf) => {
        if (!boolAccum) {
          return boolAccum;
        }
        // console.log(`comparing isequal field ${cf}, ${candidate[cf]} to ${m[cf]}`);
        return _.isEqual(c[cf], m[cf]);
      }, true);
  }
  return _.isEqual(m, c);
}
export const docxTemplateFn = (apiConfig, templateType) => path.join(apiConfig.templatesPath,
  (templateType === 'Missing' ? 'missingTemplate.docx' : 'expiredTemplate.docx'))
export const htmlTemplateFn = (apiConfig, templateType) => path.join(apiConfig.templatesPath,
  (templateType === 'Missing' ? 'missingTemplate.html' : 'expiredTemplate.html'))

