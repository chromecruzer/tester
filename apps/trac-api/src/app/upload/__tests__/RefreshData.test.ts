import {ImportIntoConsignments, ImportIntoJumpTable, SearchTable} from '@trac/postgresql';
import RefreshData from "../RefreshData";
import util from "util";
import {StatusCodes} from "http-status-codes";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

jest.mock('@trac/postgresql');
describe('RefreshData', () => {
  let mockSearchTable, module, mockUploadManager, mockPostgresClient, mockResponse, client,
    mockImportIntoConsignments, mockImportIntoJumpTable;
  beforeEach(() => {
    client = 'I am a client';
    mockSearchTable = {
      removeTable: jest.fn(),
      createTable: jest.fn(),
      fillTable: jest.fn(),
    }
    mockUploadManager = {
      markRefresh: jest.fn(),
    }
    mockPostgresClient = {
      getClient: jest.fn().mockReturnValue(Promise.resolve(client)),
      release: jest.fn()
    }
    mockResponse = {
      status: jest.fn(),
      send: jest.fn(),
    }
    mockImportIntoConsignments = {
      removeTable: jest.fn(),
      into: jest.fn(),
    }
    mockImportIntoJumpTable = {
      removeTable: jest.fn(),
      fill: jest.fn(),
    }
    jest.mocked(SearchTable, true).mockReturnValue(mockSearchTable);
    jest.mocked(ImportIntoConsignments, true).mockReturnValue(mockImportIntoConsignments);
    jest.mocked(ImportIntoJumpTable, true).mockReturnValue(mockImportIntoJumpTable);
    module = new RefreshData('test schema', mockUploadManager);
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  describe('refresh', () => {
    it('should calls the search table functions correctly', () => {
      const func = module.refresh(mockPostgresClient);
      return func({body:{user: 'Gloria Titwillow'}}, mockResponse, jest.fn())
        .then(() => {
          // console.log('func completed');
          expect(mockSearchTable.removeTable).toBeCalledWith(client);
          expect(mockSearchTable.createTable).toBeCalledWith(client);
          expect(mockSearchTable.fillTable.mock.calls).toMatchSnapshot();
        })
    });
    it('should call import consignment functions correctly', () => {
      const func = module.refresh(mockPostgresClient);
      return func({body:{user: 'Gloria Titwillow'}}, mockResponse, jest.fn())
        .then(() => {
          // console.log('func completed');
          expect(mockImportIntoConsignments.removeTable).toBeCalledWith(client);
          expect(mockImportIntoConsignments.into).toBeCalledWith(['PREM_IOLS_TOT', 'STAND_IOLS_TOT'], client);
        })
    });
    it('should call import sales jump table functions correctly', () => {
      const func = module.refresh(mockPostgresClient);
      return func({body:{user: 'Gloria Titwillow'}}, mockResponse, jest.fn())
        .then(() => {
          // console.log('func completed');
          expect(mockImportIntoJumpTable.removeTable).toBeCalledWith(client);
          expect(mockImportIntoJumpTable.fill).toBeCalledWith(client);
        })
    });
    it('should mark the refresh with the upload manager', () => {
      const func = module.refresh(mockPostgresClient);
      return func({body:{user: 'Gloria Titwillow'}}, mockResponse, jest.fn())
        .then(() => {
          // console.log('func completed');
          expect(mockUploadManager.markRefresh).toBeCalledWith('Gloria Titwillow');
        })
    });
    it('should handle the postgresql client correctly', () => {
      const func = module.refresh(mockPostgresClient);
      return func({body:{user: 'Gloria Titwillow'}}, mockResponse, jest.fn())
        .then(() => {
          // console.log('func completed');
          expect(mockPostgresClient.getClient).toBeCalled();
          expect(mockPostgresClient.release).toBeCalledWith(client);
        })
    });
    it('should return the correct response', () => {
      const func = module.refresh(mockPostgresClient);
      return func({body:{user: 'Gloria Titwillow'}}, mockResponse, jest.fn())
        .then(() => {
          // console.log('func completed');
          expect(mockResponse.status).toBeCalledWith(StatusCodes.OK);
          expect(mockResponse.send).toBeCalledWith('OK');
        })
    });
  });
});
