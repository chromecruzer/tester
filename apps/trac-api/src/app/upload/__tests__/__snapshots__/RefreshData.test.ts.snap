// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RefreshData refresh should calls the search table functions correctly 1`] = `
Array [
  Array [
    "test schema",
    "consignments",
    "I am a client",
  ],
  Array [
    "test schema",
    "customers",
    "I am a client",
  ],
  Array [
    "test schema",
    "products",
    "I am a client",
  ],
  Array [
    "test schema",
    "salesreps",
    "I am a client",
  ],
]
`;
